README.md
# pytest-dash

[Pytest][2] plugin for [Dash][1]. Pytest-dash aims to ease automated testing of dash applications and components.

[![CircleCI](https://circleci.com/gh/T4rk1n/pytest-dash.svg?style=svg)](https://circleci.com/gh/T4rk1n/pytest-dash)
[![Documentation Status](https://readthedocs.org/projects/pytest-dash/badge/?version=latest)](https://pytest-dash.readthedocs.io/en/latest/?badge=latest)
[![codecov](https://codecov.io/gh/T4rk1n/pytest-dash/branch/master/graph/badge.svg)](https://codecov.io/gh/T4rk1n/pytest-dash)

## Install

Install with pip: `$ pip install -U pytest-dash`

## Features

- Fixtures to start [Dash][1] applications and closes them in teardown.
- Declarative behavior test api in a yaml format.
- Selenium wait for wrappers.

## Documentation

For full documentation and examples, please see https://pytest-dash.readthedocs.io/en/latest/usage.html

## Bugs/Suggestions

For any bug, suggestion or request, please open a [GitHub issue](https://github.com/T4rk1n/pytest-dash/issues).

## More resources

- [Dash user guide](https://dash.plot.ly/)
- [Pytest][2]

[1]: https://github.com/plotly/dash
[2]: https://github.com/pytest-dev/pytest


CHANGELOG.md
# Changelog

Versions follow [Semantic Versioning](https://www.semver.org)

## [2.1.2] - 2019-03-18
### Fixed
- Fixed server closed assertion on non-english locales. [#70](https://github.com/T4rk1n/pytest-dash/pull/70)

## [2.1.1] - 2019-02-21
### Fixed
- Fixed python 2 super call on subclasses of BaseDashRunner [#68](https://github.com/T4rk1n/pytest-dash/pull/66)

## [2.1.0] - 2019-02-20
### Added
- Add `pytest_setup_selenium` hook [#65](https://github.com/T4rk1n/pytest-dash/pull/65)

## [2.0.0] - 2019-02-12
### Added
- Behavior tests from yaml description. [#54](https://github.com/T4rk1n/pytest-dash/pull/54)
- Plugin configuration hooks for selenium options (Currently only `webdriver`)
    - Configuration from commandline
    - Configuration from pytest.ini file.
- Added xpath wait_for wrappers.
- Added wait_for wrappers for multi elements find.
- Customizable behavior parsing thru `pytest_add_behavior` hook.

### Changed
- Application runners (dash_threaded, dash_subprocess) refactored to a context manager with own selenium driver.
- Driver argument changed to `--webdriver`

### Fixed
- dash_threaded now wait til the server has stopped in teardown.
- server runners now properly closes if there's an error in initialization. Fixes [#57](https://github.com/T4rk1n/pytest-dash/issues/57)

### Breaking Changes
#### Removed
- Removed `pytest-selenium` dependency, now incompatible, get the driver from the fixtures (eg: `dash_threaded.driver`).
- Removed `percy_snapshot` fixture.

#### Moved
- `utils.import_app` moved to `application_runners.py`

## [1.1.0] 2019-01-01
### Added
- Base exception type: `PytestDashError` [#23](https://github.com/T4rk1n/pytest-dash/pull/23)
- `DashAppLoadingError` [#23](https://github.com/T4rk1n/pytest-dash/pull/23)
  - Display the body html
  - Display console logs
  - Catch common errors early. [#33](https://github.com/T4rk1n/pytest-dash/pull/33)
  - Loop wait_for `#_dash-app-content` and retry the url. [#33](https://github.com/T4rk1n/pytest-dash/pull/33)
  - Added `start_wait_time` and `start_timeout` to `dash_threaded` [#33](https://github.com/T4rk1n/pytest-dash/pull/33)
- Add port option to `dash_threaded` and `dash_subprocess`. [#28](https://github.com/T4rk1n/pytest-dash/pull/28)
- Add `start_wait_time` option to `dash_threadred` for waiting after starting the thread, default to 1 sec. [#28](https://github.com/T4rk1n/pytest-dash/pull/28)
- Add more `wait_for` wrappers [#41](https://github.com/T4rk1n/pytest-dash/pull/41)
  - `wait_for_style_to_equal`
  - `wait_for_property_to_equal`
  - `wait_for_element_by_*`

### Fixed
- `dash_subprocess` uses `_wait_for_client_app_started` instead of polling the output, fix subprocess tests on circle #13, [#43](https://github.com/T4rk1n/pytest-dash/pull/43)

## [1.0.1] 2018-12-05
### Fixed
- Fixed `utils.import_app` imported methods not having access to imports. [#12](https://github.com/T4rk1n/pytest-dash/pull/11)

### Changed
- Syntax for `utils.import_app` changed to dot notation, same as `dash_subprocess`.

## [1.0.0] 2018-12-04

[#8](https://github.com/T4rk1n/pytest-dash/pull/8)

### Added

- Added `dash_subprocess` fixture, runs a dash app in a subprocess waitress-serve command.
- `utils.wait_for_text_to_equal`
- `utils.wait_for_element_by_css_selector`

### Removed
- `dash_app` fixture.

### Renamed
- `start_dash` fixture to `dash_threaded`

## [0.1.3] 2018-10-04
### Fixed

- Ensure the page is loaded after starting the app. [#6](https://github.com/T4rk1n/pytest-dash/pull/6)

## [0.1.2] 2018-10-04
### Fixed

- Better error for missing app in `dash_from_file` fixture. [#5](https://github.com/T4rk1n/pytest-dash/pull/5)

## [0.1.1] 2018-10-04
### Fixed

- Added fixtures usage examples to the README. [#4](https://github.com/T4rk1n/pytest-dash/pull/4)
- Fixed setup.cfg classifiers.

## [0.1.0] 2018-10-03
### Added

- Initial fixtures [#1](https://github.com/T4rk1n/pytest-dash/pull/1).
    - `start_dash`, start a dash app instance in a thread.
    - `dash_from_file`, load a py file and return the dash app.
    - `dash_app`, combine `dash_from_file` and `start_dash`.
    - `percy_snapshot`, take percy snapshot (untested)


tests/test_errors.py
# pylint: disable=missing-docstring
import dash
import pytest

from pytest_dash.errors import DashAppLoadingError, NoAppFoundError
from pytest_dash.utils import import_app


def test_invalid_start_raises(dash_threaded):
    app = dash.Dash(__name__)

    # Start the server without setting the layout.
    with pytest.raises(DashAppLoadingError):
        dash_threaded(app, start_timeout=1)


def test_no_app_found():
    with pytest.raises(NoAppFoundError):
        import_app('test_apps.invalid_app')


tests/test_apps/bad.py
print('NO app')


tests/test_integrations.py
# pylint: disable=redefined-outer-name, missing-docstring
try:
    from queue import Queue
except ImportError:
    # noinspection PyUnresolvedReferences
    from Queue import Queue

import pytest
import dash
from dash.dependencies import Output, Input
from dash.exceptions import PreventUpdate
import dash_html_components as html

# pylint: disable=unused-import
from pytest_dash.utils import \
    wait_for_text_to_equal, wait_for_element_by_css_selector, import_app


def test_dash_threaded(dash_threaded, selenium):
    app = dash.Dash(__name__)

    app.layout = html.Div([
        html.Button('click me', id='clicker'),
        html.Div(id='output')
    ])

    call_count = Queue()

    @app.callback(Output('output', 'children'), [Input('clicker', 'n_clicks')])
    def on_click(n_clicks):
        call_count.put(1)
        if n_clicks is None:
            raise PreventUpdate

        return n_clicks

    dash_threaded(app, port=8090)
    assert 'http://localhost:8090' in selenium.current_url

    clicker = wait_for_element_by_css_selector(selenium, '#clicker')

    for i in range(6):
        clicker.click()
        wait_for_text_to_equal(selenium, '#output', str(i + 1))

    assert call_count.qsize() == 7


@pytest.mark.skipif(
    'os.environ.get("CIRCLECI")', reason='Bugged because of #15'
)
def test_imported_app(dash_threaded, selenium):
    app = import_app('test_apps.simple_app')
    dash_threaded(app)

    value_input = selenium.find_element_by_id('value')
    value_input.clear()
    value_input.send_keys('Hello imported dash')

    wait_for_text_to_equal(selenium, '#out', 'Hello imported dash')


def test_subprocess(dash_subprocess, selenium):
    dash_subprocess('test_apps.simple_app', port=8080)
    assert 'http://localhost:8080' in selenium.current_url

    value_input = selenium.find_element_by_id('value')
    value_input.clear()
    value_input.send_keys('Hello dash subprocess')

    wait_for_text_to_equal(selenium, '#out', 'Hello dash subprocess')


pytest_dash/tools.py
from __future__ import print_function

import os
import threading
import time
import sys
import subprocess
import shlex
import uuid

try:
    from queue import Queue, Empty
except ImportError:
    from Queue import Queue, Empty

import pytest
import flask
import requests
import percy

from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.select import By


def _stop_server():
    stopper = flask.request.environ['werkzeug.server.shutdown']
    stopper()
    return 'stop'


def _wait_for_client_app_started(driver):
    # Wait until the react-entry-point is loaded.
    WebDriverWait(driver, 10).until(
        EC.presence_of_element_located((By.ID, '_dash-app-content')))


@pytest.fixture(scope='package')
def percy_snapshot(selenium):
    loader = percy.ResourceLoader(webdriver=selenium)
    percy_runner = percy.Runner(loader=loader)
    percy_runner.initialize_build()

    def take_snapshot(snapshot_name):
        print('Percy Snapshot {}'.format(sys.version.split(' ')[0]))
        percy_runner.snapshot(name=snapshot_name)

    yield take_snapshot

    percy_runner.finalize_build()


@pytest.fixture
def dash_threaded(selenium):
    """
    Start a local dash server in a new thread. Stop the server in teardown.

    :param selenium: A selenium fixture.
    :return:
    """

    stop_route = '/_stop-{}'.format(uuid.uuid4().hex)

    def create_app(app):

        app.server.add_url_rule(stop_route, stop_route, _stop_server)

        def run():
            app.scripts.config.serve_locally = True
            app.css.config.serve_locally = True
            app.run_server(debug=False, port=8050, threaded=True)

        t = threading.Thread(target=run)
        t.daemon = True
        t.start()
        time.sleep(3)
        selenium.get('http://localhost:8050')
        _wait_for_client_app_started(selenium)

        return app

    yield create_app

    # Stop the server in teardown
    requests.get('http://localhost:8050{}'.format(stop_route))


@pytest.fixture
def dash_from_file():
    """
    Import a dash app from a filepath, the imported file must have a Dash
    instance named `app`
    """
    yield import_app


@pytest.fixture
def dash_subprocess(selenium):
    namespace = {
        'process': None,
        'queue': Queue()
    }

    def _enqueue(out):
        for line in iter(out.readline, b''):
            namespace['queue'].put(line)
        out.close()

    def _sub(app_module):
        server_path = '{}:app.server'.format(
            '.'.join(os.path.split(app_module.replace('.py', '')))
        )

        status = None
        started = False
        is_windows = sys.platform == 'win32'

        cmd = 'waitress-serve --listen=127.0.0.1:8050 {}'.format(
            server_path
        )
        line = shlex.split(cmd, posix=not is_windows)

        # noinspection PyTypeChecker
        process = namespace['process'] = \
            subprocess.Popen(line,
                             bufsize=1,
                             creationflags=subprocess.CREATE_NEW_PROCESS_GROUP,
                             stdout=subprocess.PIPE,
                             stderr=subprocess.PIPE)

        queue_thread = threading.Thread(
            target=_enqueue,
            args=(process.stdout,),
        )
        queue_thread.daemon = True
        queue_thread.start()

        while not started and status is None:
            status = process.poll()
            try:
                out = namespace['queue'].get(timeout=.1)
                out = out.decode()
                if 'Serving on' in out:
                    started = True

            except Empty:
                pass

        if status is not None:
            _, err = process.communicate()
            print(err.decode(), file=sys.stderr)
            raise Exception('Could not start the server.')

        selenium.get('http://localhost:8050/')
        _wait_for_client_app_started(selenium)

    yield _sub

    namespace['process'].kill()
    while not namespace['process'].poll():
        time.sleep(0.01)


pytest_dash/utils.py
import runpy

from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.support.select import By

from pytest_dash.errors import NoAppFoundError


def wait_for_element_by_css_selector(driver, selector, timeout=10):
    return WebDriverWait(driver, timeout).until(
        EC.presence_of_element_located((By.CSS_SELECTOR, selector))
    )


def wait_for_text_to_equal(driver, selector, text, timeout=10):
    """

    :param driver:
    :type driver: selenium.webdriver.remote.webdriver.WebDriver
    :param selector:
    :param text:
    :param timeout:
    :return:
    """
    WebDriverWait(driver, timeout).until(
        EC.text_to_be_present_in_element((By.CSS_SELECTOR, selector), text)
    )
    assert text == driver.find_element_by_css_selector(selector).text


def import_app(app_file):
    try:
        app_module = runpy.run_module(app_file)
        app = app_module['app']
    except KeyError:
        raise NoAppFoundError(
            'No dash `app` instance was found in {}'.format(app_file)
        )
    return app


pytest_dash/application_runners.py
"""
Run dash applications with a context manager.
When exiting the context, the server will close.
"""
from __future__ import print_function
import runpy
import shlex
import subprocess
import time
import uuid
import urllib
import threading
import sys

import flask
import requests

from selenium.common.exceptions import TimeoutException
from selenium.webdriver.support.wait import WebDriverWait
from urllib.error import URLError

from pytest_dash import errors
from pytest_dash.wait_for import _wait_for_client_app_started


def _stop_server():
    stopper = flask.request.environ['werkzeug.server.shutdown']
    stopper()
    return 'stop'


def _assert_closed(driver):
    try:
        urllib.request.urlopen(driver.current_url)
    except URLError:
        return True
    return False


def _handle_error(_):
    _stop_server()


def import_app(app_file, application_name='app'):
    """
    Import a dash application from a module.
    The import path is in dot notation to the module.
    The variable named app will be returned.

    :Example:

        >>> app = import_app('my_app.app')

    Will import the application in module `app` of the package `my_app`.

    :param app_file: Path to the app (dot-separated).
    :type app_file: str
    :param application_name: The name of the dash application instance.
    :raise: pytest_dash.errors.NoAppFoundError
    :return: App from module.
    :rtype: dash.Dash
    """
    try:
        app_module = runpy.run_module(app_file)
        app = app_module[application_name]
    except KeyError:
        raise errors.NoAppFoundError(
            'No dash `app` instance was found in {}'.format(app_file)
        )
    return app


class BaseDashRunner(object):
    """Base context manager class for running applications."""

    def __init__(self, driver, keep_open=False):
        """
        :param driver: Selenium driver
        :type driver: selenium.webdriver.remote.webdriver.WebDriver
        :param keep_open: Keep the server open
        :type keep_open: bool
        """
        self.driver = driver
        self.port = 8050
        self.started = False
        self.keep_open = keep_open

    def start(self, *args, **kwargs):
        """
        Start the application.

        :param args:
        :param kwargs:
        :return:
        """
        raise NotImplementedError  # pragma: no cover

    def stop(self):
        """
        Stop the dash application.

        :return:
        """
        raise NotImplementedError  # pragma: no cover

    def __call__(self, *args, **kwargs):
        return self.start(*args, **kwargs)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.started and not self.keep_open:
            self.stop()
            try:
                WebDriverWait(self.driver, 1).until(_assert_closed)
            except TimeoutException:  # pragma: no cover
                raise errors.ServerCloseError(
                    'Could not stop server (port={})'.format(self.port)
                )

    @property
    def url(self):
        """
        The url with the port auto-formatted.

        :return: Formatted url with the port.
        """
        return 'http://localhost:{}'.format(self.port)


class DashThreaded(BaseDashRunner):
    """Runs a dash application in a thread."""

    def __init__(self, driver, keep_open=False):
        super(DashThreaded, self).__init__(driver, keep_open=keep_open)
        self.stop_route = '/_stop-{}'.format(uuid.uuid4().hex)
        self.thread = None

    # pylint: disable=arguments-differ
    def start(
            self, app, port=8050, start_wait_time=0.5, start_timeout=10,
            **kwargs
    ):
        """
        Start the threaded dash app server.

        .. seealso:: :py:func:`~.plugin.dash_threaded`

        :param app: The dash application instance.
        :type app: dash.Dash
        :param port: Port of the dash application.
        :type port: int
        :param start_wait_time: Poll rate for the server started wait
        :type start_wait_time: float
        :param start_timeout: Max time to start the server.
        :type start_timeout: float
        :param kwargs:
        :return:
        """
        app.server.add_url_rule(self.stop_route, self.stop_route, _stop_server)
        self.port = port
        app.server.errorhandler(500)(_handle_error)

        def run():
            app.scripts.config.serve_locally = True
            app.css.config.serve_locally = True
            app.run_server(debug=False, port=port, threaded=True)

        self.thread = threading.Thread(target=run)

        self.thread.daemon = True
        self.thread.start()
        try:
            _wait_for_client_app_started(
                self.driver, self.url, start_wait_time, start_timeout
            )
        except errors.DashAppLoadingError:
            self.started = self.thread.is_alive()
            raise
        else:
            self.started = True

        return app

    def stop(self):
        requests.get('{}{}'.format(self.url, self.stop_route))
        while self.thread.is_alive():
            time.sleep(0.1)


class DashSubprocess(BaseDashRunner):
    """Runs a dash application in a waitress-serve subprocess."""

    def __init__(self, driver, keep_open=False):
        super(DashSubprocess, self).__init__(driver, keep_open=keep_open)
        self.process = None

    # pylint: disable=arguments-differ
    def start(self, app_module, application_name='app', port=8050):
        """
        Start the waitress-serve process.

        .. seealso:: :py:func:`~.plugin.dash_subprocess`

        :param app_module: Dot notation path to the app file.
        :type app_module: str
        :param application_name: Variable name of the dash instance.
        :type application_name: str
        :param port: Port to serve the application.
        :type port: int
        :return:
        """
        server_path = '{}:{}.server'.format(app_module, application_name)
        self.port = port

        is_windows = sys.platform == 'win32'

        cmd = 'waitress-serve --listen=127.0.0.1:{} {}'.format(
            port, server_path
        )
        line = shlex.split(cmd, posix=not is_windows)

        # noinspection PyTypeChecker
        self.process = subprocess.Popen(
            line, stdout=subprocess.PIPE, stderr=subprocess.PIPE
        )

        url = 'http://localhost:{}/'.format(port)

        try:
            _wait_for_client_app_started(self.driver, url)
        except errors.DashAppLoadingError:
            status = self.process.poll()
            print(
                '\nDash subprocess: {} Failed with status: {}'.format(
                    cmd, status
                )
            )
            self.started = status is None
            raise
        else:
            self.started = True

    def stop(self):
        self.process.kill()
        while not self.process.poll():
            time.sleep(0.01)
        out, err = self.process.communicate()
        if out:
            print(out.decode(), file=sys.stderr)  # pragma: no cover
        if err:
            print(err.decode(), file=sys.stderr)


pytest_dash/fixtures.py
from __future__ import print_function

import pprint
import threading
import time
import sys
import subprocess
import shlex
import uuid

from selenium.common.exceptions import TimeoutException

try:
    from queue import Queue, Empty
except ImportError:
    from Queue import Queue, Empty

import pytest
import flask
import requests
import percy

from pytest_dash.errors import DashAppLoadingError
from pytest_dash.utils import wait_for_element_by_css_selector


def _stop_server():
    stopper = flask.request.environ['werkzeug.server.shutdown']
    stopper()
    return 'stop'


def _wait_for_client_app_started(driver):
    # Wait until the react-entry-point is loaded.
    try:
        wait_for_element_by_css_selector(driver, '#_dash-app-content')
    except TimeoutException:
        body = wait_for_element_by_css_selector(driver, 'body')
        logs = driver.get_log('browser')
        raise DashAppLoadingError(
            'Dash could not start: \nHTML:\n {}\n\nLOGS: {}'.format(
                body.get_property('innerHTML'),
                pprint.pformat(logs)
            )
        )


@pytest.fixture(scope='package')
def percy_snapshot(selenium):
    loader = percy.ResourceLoader(webdriver=selenium)
    percy_runner = percy.Runner(loader=loader)
    percy_runner.initialize_build()

    def take_snapshot(snapshot_name):
        print('Percy Snapshot {}'.format(sys.version.split(' ')[0]))
        percy_runner.snapshot(name=snapshot_name)

    yield take_snapshot

    percy_runner.finalize_build()


@pytest.fixture
def dash_threaded(selenium):
    """
    Start a local dash server in a new thread. Stop the server in teardown.

    :param selenium: A selenium fixture.
    :return:
    """

    stop_route = '/_stop-{}'.format(uuid.uuid4().hex)
    ns = dict(
        port=8050,
        url='http://localhost:{}'
    )

    def create_app(app, port=8050, start_wait_time=1):

        app.server.add_url_rule(stop_route, stop_route, _stop_server)
        ns['port'] = port
        ns['url'] = ns['url'].format(port)

        def run():
            app.scripts.config.serve_locally = True
            app.css.config.serve_locally = True
            app.run_server(debug=False, port=port, threaded=True)

        t = threading.Thread(target=run)
        t.daemon = True
        t.start()
        time.sleep(start_wait_time)
        selenium.get(ns['url'])
        _wait_for_client_app_started(selenium)

        return app

    yield create_app

    # Stop the server in teardown
    requests.get('{}{}'.format(ns['url'], stop_route))


@pytest.fixture
def dash_subprocess(selenium):
    namespace = {
        'process': None,
        'queue': Queue(),
        'port': 8050
    }

    def _enqueue(out):
        for line in iter(out.readline, b''):
            namespace['queue'].put(line)
        out.close()

    def _sub(app_module, server_instance='app.server', port=8050):
        server_path = '{}:{}'.format(app_module, server_instance)
        namespace['port'] = port

        status = None
        started = False
        is_windows = sys.platform == 'win32'

        cmd = 'waitress-serve --listen=127.0.0.1:{} {}'.format(
            port,
            server_path
        )
        line = shlex.split(cmd, posix=not is_windows)

        # noinspection PyTypeChecker
        process = namespace['process'] = \
            subprocess.Popen(line,
                             bufsize=1,
                             stdout=subprocess.PIPE,
                             stderr=subprocess.PIPE)

        queue_thread = threading.Thread(
            target=_enqueue,
            args=(process.stdout,),
        )
        queue_thread.daemon = True
        queue_thread.start()

        while not started and status is None:
            status = process.poll()
            try:
                out = namespace['queue'].get(timeout=.1)
                out = out.decode()
                if 'Serving on' in out:
                    started = True

            except Empty:
                pass

        if status is not None:
            _, err = process.communicate()
            print(err.decode(), file=sys.stderr)
            raise Exception('Could not start the server.')

        selenium.get('http://localhost:{}/'.format(port))
        _wait_for_client_app_started(selenium)

    yield _sub

    namespace['process'].kill()
    while not namespace['process'].poll():
        time.sleep(0.01)
